include/seqan3/alignment/matrix/detail/combined_score_and_trace_matrix.hpp

```diff
@@ -0,0 +1,337 @@
+    using score_matrix_iter_type = std::ranges::iterator_t<score_matrix_t>;
+    //!\brief The type of the trace matrix iterator.
+    using trace_matrix_iter_type = std::ranges::iterator_t<trace_matrix_t>;
+
+    // Befriend the base class.
+    template <typename, typename>
+    friend class combined_score_and_trace_matrix;
+
+    //!\brief The transform adaptor to convert the tuple from the zip view into a seqan3::detail::affine_cell_type.
+    static constexpr auto transform_to_combined_matrix_cell = std::views::transform([] (auto && tpl)
+        -> affine_cell_proxy<remove_cvref_t<decltype(tpl)>>
+    {
+        using fwd_tuple_t = decltype(tpl);
+        return affine_cell_proxy<remove_cvref_t<fwd_tuple_t>>{std::forward<fwd_tuple_t>(tpl)};
+    });
+
+    //!\brief The current iterator over the score matrix.
+    score_matrix_iter_type score_matrix_it;
+    //!\brief The current iterator over the trace matrix.
+    trace_matrix_iter_type trace_matrix_it;
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief The value type.
+    using value_type = decltype(std::declval<combined_column_type>() | transform_to_combined_matrix_cell);
+    //!\brief The reference type.
+    using reference = value_type;
+    //!\brief The pointer type.
+    using pointer = void;
+    //!\brief The difference type.
+    using difference_type = std::ptrdiff_t;
+    //!\brief The iterator concept.
+    using iterator_concept = std::input_iterator_tag;
+    //!\}
+
+    /*!\name Constructor, assignment and destructor
+     * \{
+     */
+    iterator() noexcept = default; //!< Defaulted.
+    iterator(iterator const &) noexcept = default; //!< Defaulted.
+    iterator(iterator &&) noexcept = default; //!< Defaulted.
+    iterator & operator=(iterator const &) noexcept = default; //!< Defaulted.
+    iterator & operator=(iterator &&) noexcept = default; //!< Defaulted.
+    ~iterator() = default; //!< Defaulted.
+
+    /*!\brief Initialises the iterator from the underlying matrix.
+     *
+     * \param[in] score_matrix_it \copybrief seqan3::detail::combined_score_and_trace_matrix::iterator::score_matrix_it
+     * \param[in] trace_matrix_it \copybrief seqan3::detail::combined_score_and_trace_matrix::iterator::trace_matrix_it
+     */
+    explicit iterator(score_matrix_iter_type score_matrix_it,
+                      trace_matrix_iter_type trace_matrix_it) noexcept :
+        score_matrix_it{std::move(score_matrix_it)},
+        trace_matrix_it{std::move(trace_matrix_it)}
+    {}
+    //!\}
+
+    /*!\name Element access
+     * \{
+     */
+    //!\brief Returns the range over the current column.
+    reference operator*() const
+    {
+        return views::zip(*score_matrix_it, *trace_matrix_it) | transform_to_combined_matrix_cell;
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\brief Advance the iterator to the next alignment matrix column.
+    iterator & operator++()
```

Comment:
Just a question: Whats the difference between these two ++ operator definitions and why does the first as a return value, but no input and the second one an input but no return value ? :)

Reply:
There are two increment operators: pre-increment (`++it`) and post-increment (`it++`). The latter is implemented as `iterator operator++(int)`. Now the semantic difference between those two is the following:

```cpp
auto it = rng.begin();
auto it2 = ++it; // it2 == it and both point to the element at position 1 of the rng.
auto it3 = it++; // Now it3 == it2 != it, i.e. it++ returns a copy of the iterator before the increment of it.
```
So in the last line it3 and it2 both point to the second element and it to the third element.
Since c++20 an input-iterator is not required to return a copy of the iterator pointing to the previous element, respectively a copy of the previously pointed to element. any more. This is great, since for input iterators making a copy can be difficult, because the advancing one copy will invalidate the other copy. That's why the post-increment operator is here defined as `void`.

Reply:
Thanks for this explanation ! :)